### PR TITLE
Cassandra Timestamp Store Invalidator [Invalidate the Timestamp Tables Part 2]

### PR DESCRIPTION
**Pull request by jeremyk-91**
This change:

* Adds a `TimestampStoreInvalidator` interface explaining what invalidation is and why it must be atomic
* Adds a `CassandraTimestampStoreInvalidator` (very easy with the timestamp backup runner in place)
* Some integration tests

@gsheasby @gmaretic Hopefully after the previous PR this should be easy to follow.

Do not merge until #1569 is in.

After this there are probably 2 more PRs before we properly support automated migrations:

* Automated upgrade runner (after which Cassandra migration should be supported)
* DbKVS Timestamp Store Invalidator

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1570)

### CODE REVIEW

**Comment by jeremyk-91** [comment]
(TODO: Rebase this guy on develop.)

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStoreInvalidator.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    private final CassandraTimestampBackupRunner backupRunner;
+
+    public CassandraTimestampStoreInvalidator(CassandraTimestampBackupRunner backupRunner) {
+        this.backupRunner = backupRunner;
+    }
+
+    @Override
+    @Idempotent
+    public long backupAndInvalidate() {
+        backupRunner.createTimestampTable();
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStoreInvalidator.java]: Should be renamed to `createTimestampTableIfNotExists` or `ensureTimestampTableExists`.

File: timestamp-api/src/main/java/com/palantir/timestamp/TimestampStoreInvalidator.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+public interface TimestampStoreInvalidator {
+    /**
+     * This method performs a backup of the data in an underlying timestamp bound store, and then corrupts it
+     * (if this has not already been performed). It then returns the value that has been backed up.
+     * These operations are grouped because of the requirement that as far as the timestamp bound store is concerned,
+     * this operation should be atomic. Otherwise we can have the following situation:
+     *
+     * 1. Client A reads a backup timestamp B1, preparing to invalidate and backup B1
+     * 2. Client B issues the timestamp B1 + 1
+     * 3. Client A invalidates the internal store and fast forwards a different timestamp service to B1
```
Comment (gsheasby) [inline timestamp-api/src/main/java/com/palantir/timestamp/TimestampStoreInvalidator.java]: Do we want to add multi-client tests to cover cases like this?

File: atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStoreInvalidatorIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+import com.palantir.atlasdb.containers.CassandraContainer;
+import com.palantir.atlasdb.containers.Containers;
+import com.palantir.timestamp.TimestampBoundStore;
+
+public class CassandraTimestampStoreInvalidatorIntegrationTest {
+    @ClassRule
+    public static final Containers CONTAINERS = new Containers(CassandraTimestampIntegrationTest.class)
+            .with(new CassandraContainer());
+
+    private static final long ONE_MILLION = 1000000;
```
Comment (jeremyk-91) [inline atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStoreInvalidatorIntegrationTest.java]: OK, changed.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStoreInvalidator.java (hunk 1 above)
Comment (jeremyk-91) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStoreInvalidator.java]: I like the latter, changed to that.

File: timestamp-api/src/main/java/com/palantir/timestamp/TimestampStoreInvalidator.java (hunk 2 above)
Comment (jeremyk-91) [inline timestamp-api/src/main/java/com/palantir/timestamp/TimestampStoreInvalidator.java]: Yep. I will add a few tests to C*TSInvalidatorIntegrationTest that simulate this (i.e. multiple invalidators pointing at the same KVS).

File: atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTestTools.java [hunk 4]
```diff
@@ -18,22 +18,32 @@
...
 import com.jayway.awaitility.Awaitility;
 import com.jayway.awaitility.core.ConditionTimeoutException;
+import com.palantir.common.base.Throwables;
 
 /**
  * Utilities for ETE tests
  * Created by aloro on 12/04/2016.
  */
 public final class CassandraTestTools {
+    public static final int NUM_PARALLEL_TASKS = 32;
```
Comment (gsheasby) [inline atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTestTools.java]: This should be private (and can therefore be in the same block as the other `private static final` constant).

File: atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStoreInvalidatorIntegrationTest.java [hunk 5]
```diff
@@ -90,28 +100,88 @@ public void canBackupAndRestoreTimestampTable() {
...
 
     @Test
     public void restoringTimestampTableWithNoDataIsANoOp() {
         invalidator.revalidateFromBackup();
-        checkThatWeCanReadInitialValue();
+        assertWeCanReadInitialValue();
+    }
+
+    @Test
+    public void differentInvalidatorsCanReadSameBackupBound() {
```
Comment (gsheasby) [inline atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStoreInvalidatorIntegrationTest.java]: This test looks redundant next to the below test.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBackupRunner.java [hunk 6]
```diff
@@ -88,7 +88,7 @@ public synchronized long backupExistingTimestamp() {
                             CassandraTimestampUtils.BACKUP_COLUMN_NAME,
                             Pair.create(currentBackupBound, backupValue));
             executeAndVerifyCas(client, casMap);
-            log.info("[BACKUP] Backed up the value {}", currentBackupBound);
+            log.info("[BACKUP] Backed up the value {}", currentBound);
```
Comment (gsheasby) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBackupRunner.java]: Good spot!

File: atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStoreInvalidatorIntegrationTest.java [hunk 7]
```diff
@@ -17,29 +17,41 @@
...
+    private static final long ONE_MILLION = 1_000_000;
 
     private final CassandraKeyValueService kv = CassandraKeyValueService.create(
             CassandraKeyValueServiceConfigManager.createSimpleManager(CassandraContainer.KVS_CONFIG),
             CassandraContainer.LEADER_CONFIG);
-    private final CassandraTimestampStoreInvalidator invalidator = new CassandraTimestampStoreInvalidator(
-            new CassandraTimestampBackupRunner(kv));
+    private final CassandraTimestampStoreInvalidator invalidator = CassandraTimestampStoreInvalidator.create(kv);
+
+    private static final int POOL_SIZE = 32;
```
Comment (gsheasby) [inline atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStoreInvalidatorIntegrationTest.java]: These two fields seem to be unused.

File: atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTestTools.java (hunk 4 above)
Comment (jeremyk-91) [inline atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTestTools.java]: Agreed, changed.

File: atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStoreInvalidatorIntegrationTest.java (hunk 7 above)
Comment (jeremyk-91) [inline atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStoreInvalidatorIntegrationTest.java]: Yep, you're right. This was an artifact of me experimenting with different ways of running tasks in parallel while trying to minimise the duplication between the various integration tests. Removed both fields.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBackupRunner.java (hunk 6 above)
Comment (jeremyk-91) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBackupRunner.java]: Thanks - I noticed it when reading the logs in some of the tests, and wondering why I had `Backed up the value null`.

File: atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStoreInvalidatorIntegrationTest.java (hunk 5 above)
Comment (jeremyk-91) [inline atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampStoreInvalidatorIntegrationTest.java]: Ah, right. I've removed this test; the next one is stronger.

**Comment by jeremyk-91** [comment]
OK. I'm bumping the time limit for the integration test to 10 seconds in view of that test flaking on external Circle with a timeout (which could happen given that the test containers are small).

**Comment by jeremyk-91** [comment]
Update: 10 seconds isn't enough. Looking at our data from the integration test flake at 5s observed on internal CircleCI (build 1777):

```xml
<testcase name="resilientToMultipleConcurrentBackupsAndRestores" ... time="5.684">
<failure message="com.palantir.common.exception.PalantirRuntimeException" type="com.palantir.common.exception.PalantirRuntimeException">
(stacktrace leading to timeout in executeInParallelOnExecutorService())
</failure>
</testcase>
<testcase name="backupThrowsIfBothBoundsReadable" ... time="13.474"/>
```

The `backupThrowsIfBothBoundsReadable()` test is single-threaded, doing one backup, and then another write, and then another read. I suspect that we got really unlucky with hitting a GC or Cassandra compaction, or something along those lines.

I've removed the time bound (note: CI builds will fail automatically after five minutes). I'd say at this point if we do see another failure then that more strongly suggests some kind of a deadlock in the backup runner, which is what I'm most concerned about here.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBackupRunner.java (hunk 6 above)
Comment (jboreiko) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampBackupRunner.java]: Is this going to report something funky when the db is empty?  Seems like we might want to explicitly use the backupValue as that is what we are backing up.

**Pull request merged**